### PR TITLE
SNOW-1652834: align with live connection on analyzer expression alias implementation 

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-NNNNNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

the existing implementation is incomplete as the alias map information is stored in an intermediate map which gets lost after resolving expression. this is lead to select column on joined dataframe failure because the alias map info is lost, so the column could not be found on the new joined dataframe -- see test code

in this PR, we align the implementation with live connection to persist the alias map information on the analyzer object.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_analyzer.py [hunk 1]
```diff
@@ -511,60 +611,91 @@ def table_function_expression_extractor(
...
+                    expr_str = expr_str.upper()
+                return quote_name(expr_str.strip())
+            elif isinstance(expr_str, list):
+                assert all(isinstance(e, str) for e in expr_str)
+                return ",".join([quote_name(e.strip()) for e in expr_str])
+            else:
+                raise SnowparkLocalTestingException(
+                    f"Invalid expression '{expr_str}' for UnresolvedAlias."
+                    f" Please ensure the alias name exists in the DataFrame."
+                )
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_analyzer.py:658]: Should there be an added test case for this branch?

File: src/snowflake/snowpark/mock/_analyzer.py [hunk 2]
```diff
@@ -142,22 +141,23 @@
...
         self.subquery_plans = []
+        self.generated_alias_maps = {}
         self.alias_maps_to_use = None
         self._conn = self.session._conn
 
     def analyze(
         self,
         expr: Union[Expression, NamedExpression],
-        expr_to_alias: Optional[Dict[str, str]] = None,
+        df_aliased_col_name_to_real_col_name: DefaultDict[str, Dict[str, str]] = None,
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_analyzer.py]: updated

File: src/snowflake/snowpark/mock/_analyzer.py [hunk 3]
```diff
@@ -794,6 +970,10 @@ def do_resolve_with_resolved_children(
             return MockExecutionPlan(logical_plan, self.session)
 
         if isinstance(logical_plan, MockSelectable):
+            if isinstance(logical_plan, MockSelectStatement):
+                # align with the live connection behavior where the analyzer calls
+                # logical_plan.projection_in_str to resolve the alias map
+                logical_plan.projection_in_str
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_analyzer.py:972]: yes, materializing the property resolves the intermediate alias map

File: src/snowflake/snowpark/mock/_plan.py [hunk 4]
```diff
@@ -229,6 +229,9 @@ def execution_queries(self) -> Dict[PlanQueryType, List[Query]]:
             PlanQueryType.POST_ACTIONS: self.post_actions,
         }
 
+    def add_aliases(self, to_add: Dict) -> None:
+        self.expr_to_alias = {**self.expr_to_alias, **to_add}
```
[the remark this answers is not in the document]
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_plan.py]: good point, I copied from the live code. changed to update

File: src/snowflake/snowpark/mock/_analyzer.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_analyzer.py]: this case should not happen.
I changed to assertion at the beginning and removed this error

**Review by sfc-gh-jrose — APPROVED** [review]

**Review by sfc-gh-aalam — APPROVED** [review]

**Pull request merged**